src/transfer.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  INSUFFICIENT_FUNDS_MESSAGE,
+  INVALID_TARGET_MESSAGE,
+  MAX_TOKEN_LOCK_LENGTH,
+  MIN_TOKEN_LOCK_LENGTH,
+} from './constants';
+import { Balances, Vaults, WalletAddress } from './types';
+import {
+  incrementBalance,
+  unsafeDecrementBalance,
+  walletHasSufficientBalance,
+} from './utilities';
+
+export function safeTransfer({
+  balances,
+  fromAddr,
+  toAddr,
+  qty,
+}: {
+  balances: Balances;
+  fromAddr: WalletAddress;
+  toAddr: WalletAddress;
+  qty: number;
+}): void {
+  if (qty < 0) {
+    throw new ContractError(`Quantity must be positive!`);
+  }
+
+  if (fromAddr === toAddr) {
+    throw new ContractError(INVALID_TARGET_MESSAGE);
+  }
+
+  if (balances[fromAddr] === null || isNaN(balances[fromAddr])) {
```

Comment:
Is it really necessary to use `isNaN`? I would say we're using that method to check against `null` and `undefined` values, but I'm not sure.

Also, beware there are two implementations: `window.isNaN` and `Number.isNaN`, the latter won't attempt to transform the value to a number before checking so the first one is more permissive.